Switch query parsers to use ParseField

MaineC opened the pull request:
Posting as work in progress, roughly one third of all parsers touched, more to come.

Relates to #8964

MaineC commented:
Had to add the following to turn tests to green again which went red for some of the parsers when switching to ParseFields:

    diff --git a/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java b/core/src/test/java/org/elasticsearch/index/query/Abs
    index 212dd45..10adf1a 100644
    --- a/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java
    +++ b/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java
    @@ -316,9 +316,9 @@ public abstract class AbstractQueryTestCase<QB extends AbstractQueryBuilder<QB>>
         public void testFromXContent() throws IOException {
             for (int runs = 0; runs < NUMBER_OF_TESTQUERIES; runs++) {
                 QB testQuery = createTestQueryBuilder();
    -            assertParsedQuery(testQuery.toString(), testQuery);
    +            assertParsedQuery(testQuery.toString(), testQuery, ParseFieldMatcher.STRICT);
                 for (Map.Entry<String, QB> alternateVersion : getAlternateVersions().entrySet()) {
    -                assertParsedQuery(alternateVersion.getKey(), alternateVersion.getValue());
    +                assertParsedQuery(alternateVersion.getKey(), alternateVersion.getValue(), ParseFieldMatcher.EMPTY);
                 }
             }
         }

@cbuescher does that make sense?

Reasoning behind it:

For IdsQuery here's how we document it in the reference guide:

https://www.elastic.co/guide/en/elasticsearch/reference/1.4/query-dsl-ids-query.html ... only documented parameter name is type.

Here's how we parse it: 
https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/query/IdsQueryParser.java#L73

https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/query/IdsQueryParser.java#L85

When switching to ParseFields I guessed that the documented parameter is the preferred one. This made the toXContent test fail because here we generate all three versions: 

https://github.com/elastic/elasticsearch/blob/master/core/src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTests.java#L99

cbuescher commented:
@MaineC makes sense, I think it would be great to get rid of the alternative parameter naming here, it doesn't feel consisten especially because its not documented. I'd say "type" is fine, also for the array version, but I'd wait to see what @clintongormley says about this and if we can safely remove this in 3.0. 
As for the STRICT/EMPTY ParseField flag you added, makes sense to me.

MaineC commented:
@cbuescher thanks for confirming.

Updated the preview - two tests still failing.

Made *Builder.do* protected everywhere (was public in some builders) while going through the code.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -235,10 +235,10 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
         doXArrayContent("filter", filterClauses, builder, params);
         doXArrayContent("must_not", mustNotClauses, builder, params);
         doXArrayContent("should", shouldClauses, builder, params);
-        builder.field("disable_coord", disableCoord);
```
Putting comment here for lack of beeing ablto put it at unchanged lines: just an idea, would it make sense to also introduce string constants for must/filter etc... above? I see that its not easy using them as parse fields in the current parser due to the "switch", but maybe we can change them to constant so they are all in one place, e.g. the builder?

MaineC replied:
Strange - I had the same idea seeing the code but also refrained from it due to the switch. Will look closer at actually changing it again.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/ConstantScoreQueryParser.java:
```diff
@@ -31,7 +31,7 @@
 public class ConstantScoreQueryParser implements QueryParser<ConstantScoreQueryBuilder> {
 
-    private static final ParseField INNER_QUERY_FIELD = new ParseField("filter", "query");
+    public static final ParseField INNER_QUERY_FIELD = new ParseField("filter", "query");
```
Can you also use this in L58?

MaineC replied:
? Either I updated after you added the comment or I'm not getting your comment.

cbuescher replied:
Maybe update, looks good to me now.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java:
```diff
@@ -60,23 +68,23 @@ public FuzzyQueryBuilder fromXContent(QueryParseContext parseContext) throws IOE
                     currentFieldName = parser.currentName();
                 } else {
-                    if ("term".equals(currentFieldName)) {
+                    if (parseContext.parseFieldMatcher().match(currentFieldName, TERM_FIELD)) {
```
nit: reading all this makes me think if we could get parseContext.parseFieldMatcher() once with a shorter local name at the beginning and then shorten the lines here a bit. Just for readability. I know at this points it's probably some tedious search/replace action, just throwing this in as a thought.

cbuescher replied:
After going throught the whole PR I'd probably not suggest this anymore, unless you have some nifty awk / ide search&replace magic at hand. Sorry for the noise.

MaineC replied:
Before actually making the change for all classes, what do you think of the following:

    diff --git a/core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java b/core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java
    index 85365f8..e0120c1 100644
    --- a/core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java
    +++ b/core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java
    @@ -68,23 +68,23 @@ public class FuzzyQueryParser implements QueryParser<FuzzyQueryBuilder> {
                     if (token == XContentParser.Token.FIELD_NAME) {
                         currentFieldName = parser.currentName();
                     } else {
    -                    if (parseContext.parseFieldMatcher().match(currentFieldName, TERM_FIELD)) {
    +                    if (match(currentFieldName, TERM_FIELD, parseContext)) {
                             value = parser.objectBytes();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, VALUE_FIELD)) {
    +                    } else if (match(currentFieldName, VALUE_FIELD, parseContext)) {
                             value = parser.objectBytes();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, AbstractQueryBuilder.BOOST_FIELD)) {
    +                    } else if (match(currentFieldName, AbstractQueryBuilder.BOOST_FIELD, parseContext)) {
                             boost = parser.floatValue();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, Fuzziness.FIELD)) {
    +                    } else if (match(currentFieldName, Fuzziness.FIELD, parseContext)) {
                             fuzziness = Fuzziness.parse(parser);
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, PREFIX_LENGTH_FIELD)) {
    +                    } else if (match(currentFieldName, PREFIX_LENGTH_FIELD, parseContext)) {
                             prefixLength = parser.intValue();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, MAX_EXPANSIONS_FIELD)) {
    +                    } else if (match(currentFieldName, MAX_EXPANSIONS_FIELD, parseContext)) {
                             maxExpansions = parser.intValue();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, TRANSPOSITIONS_FIELD)) {
    +                    } else if (match(currentFieldName, TRANSPOSITIONS_FIELD, parseContext)) {
                             transpositions = parser.booleanValue();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, REWRITE_FIELD)) {
    +                    } else if (match(currentFieldName, REWRITE_FIELD, parseContext)) {
                             rewrite = parser.textOrNull();
    -                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, AbstractQueryBuilder.NAME_FIELD)) {
    +                    } else if (match(currentFieldName, AbstractQueryBuilder.NAME_FIELD, parseContext)) {
                             queryName = parser.text();
                         } else {
                             throw new ParsingException(parser.getTokenLocation(), "[fuzzy] query does not support [" + currentFieldName + "]");
    diff --git a/core/src/main/java/org/elasticsearch/index/query/QueryParser.java b/core/src/main/java/org/elasticsearch/index/query/QueryParser.java
    index 0a3f6d6..4737c4e 100644
    --- a/core/src/main/java/org/elasticsearch/index/query/QueryParser.java
    +++ b/core/src/main/java/org/elasticsearch/index/query/QueryParser.java
    @@ -19,6 +19,8 @@
     
     package org.elasticsearch.index.query;
     
    +import org.elasticsearch.common.ParseField;
    +
     import java.io.IOException;
     
     /**
    @@ -48,4 +50,11 @@ public interface QueryParser<QB extends QueryBuilder<QB>> {
          * @return an empty {@link QueryBuilder} instance for this parser that can be used for deserialization
          */
         QB getBuilderPrototype();
    +    
    +    /**
    +     * @return true if the mentioned currentFieldName matches matchField
    +     * */
    +    default boolean match(String currentFieldName, ParseField matchField, QueryParseContext parseContext) {
    +       return parseContext.parseFieldMatcher().match(currentFieldName, matchField);
    +    }
     }

cbuescher replied:
Good to read, but this more or less repeats what ParseFieldMatcher.match() already does. I think it just adds another stop on the road to the actual match-implementation, which is not even in ParseFieldMatcher but in the ParseField itself. I was thinking about simple shortening like `matcher.match(currentFieldName, PARSE_FIELD)`, but I think its fine the way it is right now in the PR, not worth going trough all the files again IMHO.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/HasChildQueryParser.java:
```diff
@@ -33,7 +33,12 @@
 
-    private static final ParseField QUERY_FIELD = new ParseField("query", "filter");
+    public static final ParseField QUERY_FIELD = new ParseField("query", "filter");
+    public static final ParseField TYPE_FIELD = new ParseField("child_type", "type");
```
From the docs it seems that "type" is the preferred name here, can't judge this from the code. Can you re-check?
https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-has-child-query.html

cbuescher replied:
Interesting, the docs for HasParentQuery have "parent_type" as the documented name. Maybe another thing to sort out for more consistency.

MaineC replied:
So the docs for HasChildQuery talk about "type" only. The HasChildQueryBuilderTests.testParseFromJSON test uses "child_type". Given that as you mention HasParentQuery uses parent_type I think it would be best to stick with child_type.

After looking at the examples in our query DSL docs fixing the examples in the docs sounds like an additional PR to me.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/HasParentQueryParser.java:
```diff
@@ -30,9 +30,10 @@
-    private static final ParseField SCORE_FIELD = new ParseField("score_mode").withAllDeprecated("score");
-    private static final ParseField TYPE_FIELD = new ParseField("parent_type", "type");
+    public static final ParseField QUERY_FIELD = new ParseField("query", "filter");
+    public static final ParseField SCORE_MODE_FIELD = new ParseField("score_mode").withAllDeprecated("score");
```
Can you switch this around and use the preferred name as first constructor argument? This way it looks like there's something special with this field, which I guess its not.

MaineC replied:
Hmm - I believe didn't change that line, just checked, this https://github.com/elastic/elasticsearch/commit/d2e53e0e0c1c20f1be178a1da8bc76c9e5993cf8#diff-838f43c5421ea0e3f74f4548fddba923R50 is where it comes from.

So - looked closer at the implementation of ParseField. This thing is weird: When switching to the version you proposed this would change the behaviour at least for getPreferredName() which we use for printing the json:


    System.out.println((new ParseField("score_mode").withAllDeprecated("score"))getPreferredName());

Gives score_mode, while

    System.out.println((new ParseField("score", "score_mode")).getPreferredName());

gives score. I'm confused. Looking further what we actually want here.

MaineC replied:
For reference after chatting f2f: Before switching to ParseFields we did output "score" as fieldname in the builder which if continuing to use this syntax going forward we would not. Simplifying to use new ParseField("score", "score_mode") instead.

MaineC replied:
Looking further here after comparing to the docs:

The lower query here https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-has-parent-query.html parses well with the original code here https://github.com/elastic/elasticsearch/blob/148265bd164cd5a614cd020fb480d5974f523d81/core/src/main/java/org/elasticsearch/index/query/HasParentQueryParser.java#L34 but not when switching to SCORE_FIELD = new ParseField("score", "score_mode"). So apparently there's something I don't quite understand yet with how ParseField works.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/IdsQueryParser.java:
```diff
@@ -32,6 +33,11 @@
 public class IdsQueryParser implements QueryParser<IdsQueryBuilder> {
 
+    public static final ParseField TYPES_FIELD = new ParseField("type", "types", "_type");
+    // TODO weird: combinations types/type and _type/type were in the original code, but not _type/types/type??
```
To me it looks like type/types is for the array version, _type/type for the single key-value version. So I'm not sure if "types" should count as deprecated name.

MaineC replied:
Ah - makes much more sense now. Will change.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/IndicesQueryParser.java:
```diff
@@ -32,8 +32,11 @@
+    public static final ParseField QUERY_FIELD = new ParseField("query");
+    public static final ParseField NO_MATCH_QUERY = new ParseField("no_match_query");
+    public static final ParseField INDEX_FIELD = new ParseField("index");
+    public static final ParseField INDICES_FIELD = new ParseField("index");
```
Where did "indices" go?

MaineC replied:
Thanks for catching that.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MatchQueryParser.java:
```diff
@@ -82,44 +100,44 @@ public MatchQueryBuilder fromXContent(QueryParseContext parseContext) throws IOE
+                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, TYPE_FIELD)) {
                         String tStr = parser.text();
-                        if ("boolean".equals(tStr)) {
+                        if (parseContext.parseFieldMatcher().match(tStr, BOOLEAN_FIELD)) {
```
While using ParseField also for values is useful, I'm wondering if it might be confusing on the long run. I would at least rename BOOLEAN_FIELD and the following to something else, since technically they are not fields. No strong opinion though.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/NestedQueryParser.java:
```diff
@@ -30,6 +31,10 @@
 
     private static final NestedQueryBuilder PROTOTYPE = new NestedQueryBuilder("", EmptyQueryBuilder.PROTOTYPE);
+    public static final ParseField SCORE_MODE_FIELD = new ParseField("score_mode");
+    public static final ParseField PATH_FIELD = new ParseField("path_field");
```
This was called "path" before.

cbuescher replied:
I just realized with this, tying the field names we use in doXContent() to the parsers is great, but will hide accidental renames like this because at least for the generic query builder tests we the always use matching fieldnames. Maybe we need at least one test case per query that tries parsing a completly specified query from a query string using the prefered names. wdyt?

MaineC replied:
+1 to this kind of test. I've already seen it with some queries.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/PrefixQueryParser.java:
```diff
@@ -30,7 +30,9 @@
 
-    private static final ParseField NAME_FIELD = new ParseField("_name").withAllDeprecated("query name is not supported in short version of prefix query");
+    public static final ParseField NAME_FIELD = new ParseField("_name").withAllDeprecated("query name is not supported in short version of prefix query");
+    public static final ParseField PREFIX_FIELD = new ParseField("value", "prefix");
```
I'm not sure if this might be a case where both names are allowed, at least the doc state both version, not mentioning any deprecation.

MaineC replied:
Putting myself into the user's shoes: I personally would prefer having one and only one non-deprecated way of specifying something - makes it easier to compare different implementations without worrying what the subtle difference is when using either option. Not sure which version to keep here though.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/PrefixQueryParser.java:
```diff
@@ -30,7 +30,9 @@
 public class PrefixQueryParser implements QueryParser<PrefixQueryBuilder> {
 
-    private static final ParseField NAME_FIELD = new ParseField("_name").withAllDeprecated("query name is not supported in short version of prefix query");
+    public static final ParseField NAME_FIELD = new ParseField("_name").withAllDeprecated("query name is not supported in short version of prefix query");
```
Took me a while to understand why this is here: so I think NAME_FIELD so far was used only in the short version of prefix query, which is: `{ "prefix" : { "user" : "ki" }}`, so by deprecating "_name" it would throw exception in STRICT mode. So by using AbstractQueryBuilder.NAME_FIELD in the allowed position and this one in the deprecated short version position this should be okay. Can you rename it to something like DEPRECATED_NAME_FIELD or something though? Maybe somebody else should look at this also.

MaineC replied:
+1 on renaming.

According to "git praise" @javanna you originally introduced the NAME_FIELD - not sure if you had a closer look at this change when going through the PR last time you commented.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/PrefixQueryParser.java:
```diff
@@ -74,7 +76,7 @@ public PrefixQueryBuilder fromXContent(QueryParseContext parseContext) throws IO
                 }
             } else {
-                if (parseContext.parseFieldMatcher().match(currentFieldName, NAME_FIELD)) {
+                if (parseContext.parseFieldMatcher().match(currentFieldName, NAME_FIELD)) {   // this looks weird given that we already parse the _name field above?!
```
Re: your comment: left my thoughts on this above, maybe its time to get rid of this version alltogether? Not my call though.

javanna replied:
we can remove this stuff: https://github.com/elastic/elasticsearch/pull/12966 it's deprecated since 2.0, no reason to keep the deprecated version of it.

cbuescher replied:
Great!

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/SpanNotQueryParser.java:
```diff
@@ -30,6 +31,12 @@
  */
 public class SpanNotQueryParser implements QueryParser<SpanNotQueryBuilder> {
 
+    public static final ParseField POST_FIELD = new ParseField("post");
```
No query builder here that can use the parse fields as well?

MaineC replied:
Good catch.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/TermQueryParser.java:
```diff
@@ -32,6 +32,8 @@
 
     private static final ParseField NAME_FIELD = new ParseField("_name").withAllDeprecated("query name is not supported in short version of term query");
     private static final ParseField BOOST_FIELD = new ParseField("boost").withAllDeprecated("boost is not supported in short version of term query");
```
Maybe rename NAME_FIELD and BOOST_FIELD here as well, since they are only here to support the deprecated short version.

MaineC replied:
Or remove like we did above? (Deprecation happened in 2.x)

cbuescher commented:
@MaineC did a first check on all the changes, this is great. Since it is a huge PR I left a couple of comments, mostly minor. One thing that came to mind when seeing a change where apparently by accident you changes a field name in NestedQueryParser to "path_field": using the parse fields names both while rendering the queries in doXContent as well as in the parser will make accidental renamings like this hard to spot by the current parser tests, unless we have some case where we send the whole query via json string. Maybe we need additional tests for this, but maybe also outside this PR. Thoughts?

MaineC commented:
+1 on having "real" json tests in addition to the roundtripping thing.

I think one could see this as related to the stuff I did for the java api docs: What we really want is to have codified checks of the examples that are mentioned in the Query DSL docs so we don't accidentally rename parameters. Should relate to what @dadoonet is looking into at the moment.

MaineC commented:
Oh - forgot: Thanks to you both, @cbuescher and @javanna for the super speedy review and helpful comments!

MaineC commented:
Just a heads up: I ran an experiment extracting the code snippets we have in the docs and tossing them into a unit test for parsing. Though the current state isn't pretty at all it surfaced some inconsistencies compared to the actual code in our docs already. Working on isolating them, getting the "half-way-generated-half-way-cobbled-together" test into a better state right now.

Maybe one more data point supporting that the examples we include in our docs should be put under test (though fixing docs usually is the lowest entry bar for submitting a PR I can think of ;) )

clintongormley commented:
>  I think it would be great to get rid of the alternative parameter naming here, it doesn't feel consisten especially because its not documented. I'd say "type" is fine, also for the array version,

++

MaineC commented:
Update: I believe I addressed all comments.

I found the parameter renaming issues you found scary enough to go through the json query dsl reference guide, parse the examples therein into a separate test class to see if everything remains parseable:

https://github.com/elastic/elasticsearch/pull/14249/files#diff-2528bf3a5b01f96751f6fed1ee8f8189R26

No surprise: There were a few typos in the docs as well as a few things that are documented but do not work. However I also found another parameter name that I incidentally renamed. 

To me that's another argument to add more tests to the suit that start with the json representation of the query. However the class linked above looks ugly enough to me that I'd like to throw it away rather sooner than later. For illustration I moved one of the dsl reference guide example queries from above over to the corresponding unit test here:

https://github.com/elastic/elasticsearch/pull/14249/files#diff-262a49e710e2ff18f41b578be2e0fdf7R250

(essentially copying the approach taken here: https://github.com/elastic/elasticsearch/blob/09d2b4130003191631857d5b8a59c22c84ea8fbb/core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java#L200 )

Caveat: Though the change is simple enough for each query, doing it for all will take more than five minutes. So a second look confirming that this is really where we want to go is highly welcome.

MaineC commented:
Pulled json parsing tests into AbstractQueryTestCase (including the logging method that was part of HasChildQueryBuilderTests), added json parsing tests based on the examples in the reference guide to each query builder, found (and fixed) yet another subtle change in json generation I introduced in the commit introducing ParseFields that is part of this PR.

Ran gradle check successfully.

@martijnvg a second pair of eyes as discussed checking nothing broke would be greatly appreciated.

@cbuescher I added two methods to the AbstractQueryTestCase that I believe are in need for a better name - if you have any ideas, let me know :)

cbuescher commented:
@MaineC Isabel must have switched @martijnvg and @Mpdreamz, I think the earlier comment was directed at @Mpdreamz in response to your recent conversation.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -858,4 +858,47 @@ protected MultiTermVectorsResponse executeMultiTermVectors(MultiTermVectorsReque
+        XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
+        query.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        System.out.println(builder.string());
+        logger.info(msg(json, builder.string()));
```
Another idea about the logging here. While this helps at debugging, I think we should only log the query diff when whitespace-normalized queries are not equal (when the following assertion trips), maybe in a finally block?

cbuescher commented:
@MaineC Great +1 for adding at the "reference" json query to the test cases. I left some minor comments. 
Besides that I'm on the fence about the additional test assertions added with the "assertQueryParsedFromJson()" method to each test case. The selection which options are tested there are not entirely clear to me and could seem a bit random for someone casually reading the test. I think most of the parsing should be tested by the existing randomized `testFromXContent()` or am I missing something? On the other hand I understand that it's probably good to have some "controlled" test cases for the parsers.

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -858,4 +858,47 @@ protected MultiTermVectorsResponse executeMultiTermVectors(MultiTermVectorsReque
+        // now assert that we actually generate the same JSON
+        XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
+        query.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        System.out.println(builder.string());
```
leftover?

MaineC commented:
> @MaineC Isabel must have switched @martijnvg and @Mpdreamz, I think the earlier comment was 
> directed at @Mpdreamz in response to your recent conversation.

@cbuescher yeah - correct - we have too many Martijns in the autocomplete ;)

MaineC commented:
Cleaned and squashed. @cbuescher I think this is ready for another look.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MatchQueryParser.java:
```diff
@@ -45,11 +62,9 @@ public MatchQueryBuilder fromXContent(QueryParseContext parseContext) throws IOE
             type = MatchQuery.Type.PHRASE;
-        } else if ("match_phrase_prefix".equals(parser.currentName()) || "matchPhrasePrefix".equals(parser.currentName()) ||
-                "text_phrase_prefix".equals(parser.currentName()) || "textPhrasePrefix".equals(parser.currentName())) {
+        } else if (parseContext.parseFieldMatcher().match(parser.currentName(), MATCH_PHRASE_FIELD)) {
```
Can you have a quick look again if this should be MATCH_PHRASE_PREFIX_FIELD instead? In which case it would also be good to catch this in tests if it is wrong.

MaineC replied:
Yes, it has to be MATCH_PHRASE_PREFIX_FIELD instead.

Thinking about the tests: This particular option points to an interesting problem - some of the query options exclude one another. So if over time we want to extend those tests to include all/many more/insert your own preference which I think we should, I need to think about a way to allow for generating more than one json string - correct?

cbuescher replied:
I think just adding one special test case for this edge case should be okay for now.

cbuescher commented:
@MaineC I went thorugh the PR again, found one potential issue to look into but other than that seems good to me.

MaineC commented:
Thought a bit about how one could support checking more than one json, here's one proposal:

    diff --git a/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java b/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java
    index 35d883d..dbc9baa 100644
    --- a/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java
    +++ b/core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java
    @@ -869,19 +869,9 @@ public abstract class AbstractQueryTestCase<QB extends AbstractQueryBuilder<QB>>
             throw new UnsupportedOperationException("this test can't handle MultiTermVector requests");
         }
     
    -    public void testParseFromJSON() throws IOException {
    -        String json = createParseableQueryJson();
    -        QB query = (QB) parseQuery(json);
    -        assertQueryParsedFromJson(json, query);
    -
    -        // now assert that we actually generate the same JSON
    -        XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
    -        query.toXContent(builder, ToXContent.EMPTY_PARAMS);
    -        assertEquals(msg(json, builder.string()), json.replaceAll("\\s+",""), builder.string().replaceAll("\\s+",""));
    -    }    
    -
         /**
    -     * Override this method to provide a valid json string representing the query under test.
    +     * Call this method to check a valid json string representing the query under test against
    +     * it's generated json.
          *
          * Note: By the time of this writing (Nov 2015) all queries are taken from the query dsl
          * reference docs mirroring examples there. Here's how the queries were generated:
    @@ -895,10 +885,15 @@ public abstract class AbstractQueryTestCase<QB extends AbstractQueryBuilder<QB>>
          * <li> By now the roundtrip check for the json should be happy.
          * </ul>
          **/
    -    protected abstract String createParseableQueryJson();
    -    
    -    /** Add any checks on the query builder parsed from the json above here. */
    -    protected abstract void assertQueryParsedFromJson(String json, QB parsed); // I need a better name.
    +    public void checkGeneratedJson(String expected, QB source) throws IOException {
    +        // now assert that we actually generate the same JSON
    +        XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
    +        source.toXContent(builder, ToXContent.EMPTY_PARAMS);
    +        assertEquals(
    +                msg(expected, builder.string()),
    +                expected.replaceAll("\\s+",""),
    +                builder.string().replaceAll("\\s+",""));
    +    }    
     
         private String msg(String left, String right) {
             int size = Math.min(left.length(), right.length());
    diff --git a/core/src/test/java/org/elasticsearch/index/query/MatchQueryBuilderTests.java b/core/src/test/java/org/elasticsearch/index/query/MatchQueryBuilderTests.java
    index d8aeb43..3f8ade2 100644
    --- a/core/src/test/java/org/elasticsearch/index/query/MatchQueryBuilderTests.java
    +++ b/core/src/test/java/org/elasticsearch/index/query/MatchQueryBuilderTests.java
    @@ -236,9 +236,8 @@ public class MatchQueryBuilderTests extends AbstractQueryTestCase<MatchQueryBuil
             }
         }
     
    -    @Override
    -    protected String createParseableQueryJson() {
    -        return "{\n" + 
    +    public void testSimpleMatchQuery() throws IOException {
    +        String json = "{\n" + 
                     "  \"match\" : {\n" + 
                     "    \"message\" : {\n" + 
                     "      \"query\" : \"to be or not to be\",\n" + 
    @@ -254,11 +253,10 @@ public class MatchQueryBuilderTests extends AbstractQueryTestCase<MatchQueryBuil
                     "    }\n" + 
                     "  }\n" + 
                     "}";
    -    }
    +        MatchQueryBuilder qb = (MatchQueryBuilder) parseQuery(json);
    +        checkGeneratedJson(json, qb);
     
    -    @Override
    -    protected void assertQueryParsedFromJson(String json, MatchQueryBuilder parsed) {
    -        assertEquals(json, "to be or not to be", parsed.value());
    -        assertEquals(json, Operator.AND, parsed.operator());
    +        assertEquals(json, "to be or not to be", qb.value());
    +        assertEquals(json, Operator.AND, qb.operator());
         }
     }

cbuescher commented:
@MaineC the diff is a bit hard to read, but if I understand correctly you are moving most of the test (json string setup and assertions) into the test class and use `checkGeneratedJson` only as a common helper (might even be static then?). I think that is a good idea since it leaves it up to each individual query builder test how and how often to make use of it and it gets rid of the `assertQueryParsedFromJson` which felt a bit awkward anyway.

MaineC commented:
You did understand correctly. Will follow along this path then.

MaineC commented:
@cbuescher Thanks for the feedback. Here's what I did:

- fixed the Match Query parser parameter typo
- moved all json parsing test logic to the respective query test class leaving only the json comparison methods as static helper methods in AbstractQueryTestCase
- add a test for the Match Query parameter name.

I left the commits separate to make reviewing at least a little bit easier. Tests are still running on my local box while I'm typing this.

MaineC commented:
Renamed has-child-query:child_type query parameter back to "type" as per https://github.com/elastic/elasticsearch/pull/14800#discussion_r45340994

cbuescher commented:
Thanks, LGTM

clintongormley commented:
Nice, @MaineC! Can https://github.com/elastic/elasticsearch/issues/8964 be closed?

MaineC commented:
@clintongormley #8964 talks about replacing all parameter parsing, this PR only switched over the stuff related to queries, @colings86 is making similar changes while he's touching aggregations - but this still leaves us with IIRC roughly 100 classes to move over last time I counted.
